feat(governance): Heavy revisions and clarifications to community roles, processes

robzienert opened the pull request:
Covers most of the scope of #157.

Things of note:

1. I originally wrote it with language for `CODEOWNERS` files, that's been adjusted for the time being to continue use of `membership.yml`.
1. I'm setting up some new mailing lists; I'd like to see if Google Groups can become a decent open channel that isn't Slack for contributors to talk.
1. I heavily adjusted the requirements for Member (which is new), Reviewer and Approver.
1. I changed the responsibilities of a SIG Lead quite a lot.

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
-* Matt Duftler (Google)
-* Peter Stout (Netflix)
-* Erin Kidwell (Netflix)
-* Andrew Philips (Google)
```
SoT defined here: https://github.com/spinnaker/governance/blob/master/committee-steering/README.md

ajordens commented:
I'm a particularly big +1 on the idea of per-project approvers/reviewers.

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
+* Subscribed to [spinnaker-dev@googlegroups.com](TODO)
+* Sponsored by 2 Reviewers. **Note the following requirements for sponsors**:
+  * Sponsors must have close interactions with the prospective member - e.g. code/design/proposal review, coordinating issues, etc.
+  * Sponsors must be Reviewers or Approvers in any project in the [Spinnaker org](https://github.com/spinnaker).
```
Why "may"? That makes it sound like it's optional.

ajordens replied:
Read a bit strange to me, that's all ... perhaps `must be Reviewers or Approvers in at least one Spinnaker project.`

Either way, intent is fine with me.

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
+* An active member of the community.
+* Reviewed at least 5 [substantial](committee-technical-oversight/substantial-contributions.md) PRs to the codebase.
+* Knowledgable about the codebase (exhibited via discussion, issues, thoughtful PR review, etc.)
+* Sponsored by 1 Approver of the project
```
I was dropping it because, since it is going to be per-project, we likely won't have that many Approvers. We can change it to be 2 Approvers for any project, though, for the time being?

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
-* All Reviewer duties
+* Reviewer of the codebase for at least 3 months.
+* Reviewed or merged at least 20 PRs to the codebase.
+* Nominated by at least 2 other Approvers or a SIG Lead.
```
Yeah, I'd like to (separately) take a review of what SIGs exist and map areas of functionality to a SIG. If there's no owning SIG, then I'd like to either create one, or explicitly loop it into a SIG (preferred, given the community size).

ajordens replied:
Agree, ideally not through minting of new SIGs.

Thematically we're missing a destination for the following:
* GCS/GCE/GAE
* Halyard
* Notifications/General Orchestration
* Docker Registry
* Spin CLI

Few Observations:
* security questions (fielded by sig/platform or sig/operations?)
* ui issues (fielded by sig/ui-ux or cloud-provider?)
* Spinnaker Monitoring (deprecate and take the plugin to sig/platform or sig/operations?)

brian-armory replied:
+1 to having a mapping. That would really help out the Docs SIG so that we know who to ping to review issues/PRs.

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
+Reviewers are able to review code for quality and correctness on some part of the project. They are knowledgable about both the codebase and software engineering principles. Reviewership is required before becoming an Approver (except under specific conditions; see Approver).
+
+**Defined by**: _reviewers_ entry in the [membership.yml](membership.yml) file.
 
-Duties
+Reviewer status is scoped to a part of the codebase.
```
I was thinking that `membership.yml` in the governance repo would go away, but I'm happy to change this PR to be more aligned with using `membership.yml`.

I added the "defined by" stuff because it was part of the k8s stuff, but I liked that there was a clear definition of where people can find where the source of truth is for this information. We can link to it some other way, though.

robzienert commented on governance.md:
```diff
@@ -1,133 +1,233 @@
+* Reviewed at least 5 [substantial](committee-technical-oversight/substantial-contributions.md) PRs to the codebase.
+* Knowledgable about the codebase (exhibited via discussion, issues, thoughtful PR review, etc.)
+* Sponsored by 1 Approver of the project
+  * With no objections from other Approvers
```
Sure, we can remove vetos, but I felt like it would be worth having if we're looking to get rid of the TOC voting on each membership / promotion. Let the community do things, and if someone has a problem with an Approver / Reviewer promotion, they'd have the ability to affect it.

ezimanyi replied:
I don't have a strong opinion one way or another, but I guess one question with vetos is how long you leave the request open...if someone gets a sponsor and opens the PR we'll probably merge as soon as it meets the requirements.  Should we leave these open for a while to wait for objections?  Or remove the status if someone objects within a certain amount of time.  Maybe I'm being pedantic, just wanted to think through what this would look like.

robzienert commented:
Updated with the following changes:

- Typo fixes.
- Removed per-project distinctions.
- Loosened criteria for nominations / approvals (further reducing dependency on TOC for role changes)

demobox commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+| Member | Active contributor in the community | Sponsored by 2 Reviewers. | Spinnaker GitHub org member |
+| Reviewer | Review contributions from other members | History of review and authorship in a project | [membership.yml](membership.yml) file |
+| Approver | Approve incoming contributions | Highly experienced and active reviewer / contributor to a project | [membership.yml](membership.yml) file |
+| SIG Lead | Facilitate SIG organization, set direction and priorities for a SIG | Demonstrated responsibility and excellent judgment for the domain | [membership.yml](membership.yml) file | 
```
```suggestion
| SIG Lead | Facilitate SIG organization, set direction and priorities for a SIG | Demonstrated responsibility and excellent judgement for the domain | [membership.yml](membership.yml) file | 
```

devorbitus commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+* Work with other SIG Leads to maintain the project's overall health and success.
+* Triage inbound issues related to the SIG.
+
+**Note**: SIG Leads who are not Approvers are expected to proactively perform code reviews and sumit contributions, working towards becoming an Approver.
```
FYI: typo on submit `perform code reviews and sumit contributions`

devorbitus replied:
```suggestion
**Note**: SIG Leads who are not Approvers are expected to proactively perform code reviews and submit contributions, working towards becoming an Approver.
```

plumpy commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+  * Authoring or reviewing PRs on GitHub
+  * Filing or commenting on Issues in GitHub
+  * Contributing to SIG or community discussions (e.g. meetings, Slack, email discussions)
+* Subscribed to [spinnaker-dev@googlegroups.com](TODO)
```
`dev@spinnaker.io`?

plumpy commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+
+In some specific cases, Members can directly climb to Approver. In these cases, an issue must be made in the governance repository and voted on unanimously by the Technical Oversight Committee. Scenarios in which an Approver role can be expedited are:
+
+* Subproject ownership transfers.
```
...what does this mean? I suppose you're specifically trying to call out the `spin` case here? It kinda felt like a one-off case, and maybe it's just enough to say it's up to the TOC's discretion in rare circumstances and with their unanimous approval?

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+Reviewers are able to review code for quality and correctness on some part of the project. They are knowledgable about both the codebase and software engineering principles. Reviewership is required before becoming an Approver (except under specific conditions; see Approver).
 
-If you’d like to nominate yourself as a Reviewer, please create an [Issue in the governance repository](https://github.com/spinnaker/governance/issues). 
+Reviewer status is scoped to a part of the codebase.
```
Did we end up deciding on repo-level status?  Or is the plan that for now it will apply to the whole codebase but we might split out parts as needed (ex: kleat/spin/halyard)?

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,229 @@
-* Merge code into master
-* All Reviewer duties
+* Reviewer of the codebase for at least 3 months.
+* Reviewed or merged at least 20 PRs to the codebase.
```
I'm guessing "merged" here means "authored", people won't be able to actually merge anything before getting Approver status...if so I'd suggest changing to "authored".

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,233 @@
+## Documentation Contributors
 
-Election to the SC requires a majority of the SC. If a person is up for reelection they are not allowed to vote for their reelection. 
+The role definitions above were written optimized for code contributors. Documentation contributors will not be held to the same technical requirements. Instead, we look for writers who can express concepts clearly and simply.
```
Yeah, I wonder if this is already assumed by per-project permissions.

One thing that's not clear to me above---are we saying that docs reviewers/approvers don't have to follow the above requirements?  Or should they need to follow the same requirements but considering docs changes?

Maybe the Docs SIG can weigh in, but in reading the above requirements I think we can apply them to docs contributsions as well.  So to be a docs reviewer, you would need to be knowledgable about the docs codebase, have reviewed 5 substantial docs PRs (ex: not typos).  And then similarly for approver, where we apply the requirements in the context of docs changes.

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,229 @@
+
+A core principle in maintaining a healthy community is encouraging active participation. It is inevitable that peoples' focuses will change over time and they are not expected to be actively contributing forever. However, being a member of the Spinnaker GitHub organization comes with an elevated set of permissions. These capabilities should not be used by those who are unfamiliar with the current ...
+
+Members with an extended period away from the project with no activity will be removed from the Spinnaker GitHub organization and will be required to go through the membership process again after re-familiarizing themselves with the current state. Any time restrictions on a role will be waived when rejoining the organization.
```
To make sure I understand...you'd still need to hit the minimum number of PRs (reviewed or authored) when coming back?  Or if they made some contributions and re-familarized themselves would the old ones then still count?

This might not be super important to define right now as it's probably rare, just one thing I wasn't sure of from the phrasing.

robzienert replied:
Clarified that prior contributions will not be counted.

robzienert commented:
Updated with a lot of the feedback applied. No subproject shenanigans, as that will take a substantial overhaul of tooling which none of us have time to perform, but the rest is all there.

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,226 @@
-[Current Approvers](https://github.com/spinnaker/community/blob/master/membership.yml)
+#### Expedited Approvership
+
+In rare circumstances, Members can directly climb to Approver without going through Reviewer. In these cases, unanimous approval must be reached by the TOC. Examples of scenarios that have caused an expited Approver role have been:
```
```suggestion
In rare circumstances, Members can directly climb to Approver without going through Reviewer. In these cases, unanimous approval must be reached by the TOC. Examples of scenarios that have caused an expedited Approver role have been:
```

ezimanyi commented on governance.md:
```diff
@@ -1,133 +1,226 @@
+## Documentation Contributors
 
-Election to the SC requires a majority of the SC. If a person is up for reelection they are not allowed to vote for their reelection. 
+The role definitions above were written optimized for code contributors. Documentation contributors will not be held to the same technical requirements. Instead, we look for writers who can express concepts clearly and simply. Documnetation contributors still follow the same Member, Reviewer, Approver progression ladder.
```
```suggestion
The role definitions above were written optimized for code contributors. Documentation contributors will not be held to the same technical requirements. Instead, we look for writers who can express concepts clearly and simply. Documentation contributors still follow the same Member, Reviewer, Approver progression ladder.
```
